src/main/java/com/mwt/explorers/EpsilonGreedyExplorer.java

```diff
@@ -45,22 +45,23 @@ public void updatePolicy(Policy<T> newPolicy) {
     this.defaultPolicy = newPolicy;
   }
 
-  public ChosenAction chooseAction(long saltedSeed, T context) {
+  public DecisionTuple chooseAction(long saltedSeed, T context) {
+    int numActionsForContext = getNumActions(context);
     Random random = new Random(saltedSeed);
```

Comment:
I haven't got to testing yet but it definitely makes correctness verification a lot easier. The C++ version is using a built-in random generator I believe, so we'll need to converge on one, possibly something already in VW since we're using the hash there.